libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt

```diff
@@ -315,6 +345,18 @@ class MapboxNavigation(
         ThreadController.cancelAllUICoroutines()
     }
 
+    /**
+     * Set [RerouteController]. By default uses [MapboxRerouteController]
+     */
+    fun setRerouteController(rerouteController: RerouteController) {
+        this.rerouteController = rerouteController
+    }
+
+    /**
+     * Get [RerouteController]
+     */
+    fun getRerouteController(): RerouteController = rerouteController
```

Comment:
yep, it was, but it was replaced by getter/setter to has the same behaviour as other properties

Reply:
The reroute and arrival controllers should have the same API surface. Arrival currently exposes: `attachArrivalController` and `removeArrivalController`. The reroute controller should expose the same so that it's consistent and additionally the getter so that we can listen to state changes.

We can also rename `attachArrivalController` to `setArrivalController` if that's better.

Reply:
If we rename it to `setArrivalController`, I guess we can make both reroute and arrival controllers properties.

Reply:
at that particular case it's not attach/detach, it's set/get that is little bit different. 
Seems that `ArrivalController` is more `Arrival Controller + Observer`, so may be we need a new name for it

Reply:
The observer part is only internal, developers won't know about it. I think both controllers should have a consistent public API.

Reply:
they have different signature, if they have similar API it will confuse developer

Reply:
@RingerJK why can't we have:
```
attachArrivalController
removeArrivalController
attachRouteController
getRerouteController
removeRerouteController
```
or
```
var arrivalController
removeArrivalController

var rerouteController
removeRerouteController
```

Reply:
@LukasPaczos not really fits for API:
1. we don't have getter for `ArrivalController`;
2. we don't have `removeRerouteController` at all, for current version we always have `rerouteController`, so method might have name `replaceReroutControllerByDefault`, but I wouldn't like expose public API for now(IMO better do it step by step)
3. if suffix `*Controller` confuses, `RerouteController` might be renamed to `RerouteManager/Handler/Maintainer/...` or smth else

Reply:
Some pros & cons 👀 

- properties you get dot notation from Kotlin
- properties or setter / getter nomenclature IMO should be only used for POJOs / DTOs
- properties or setter / getter don't allow to give domain context / naming

Reply:
Both arrival and reroute controller (regardless of how they're called) can be used to delegate a specific action to a different piece of code, whether that's arrival calculation or reroute handling, or anything else in the future. In my opinion, they should have a consistent prefix so that the behavior is immediately clear and it's easy to discover.

Reply:
@LukasPaczos I see what you mean, and agree, that similar things should have similar api(just like we're doing with observers and _register_/_unregisterObserver_). 
At this case API smooth different: when we call `attachArrivalController` it means that here's no arrival controller and we attach new one. For RerouteController we have different strategy, as we already have one we just replace it.

Reply:
> At this case API smooth different: when we call attachArrivalController it means that here's no arrival controller and we attach new one. For RerouteController we have different strategy, as we already have one we just replace it.

@RingerJK we are also attaching the default arrival controller ourselves and the exposed method replaces it. Check out https://github.com/mapbox/mapbox-navigation-android/blob/589c5a45a7cccddd6363b436699ccc525389ea2e/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L203.

Reply:
@LukasPaczos yeah, I've seen it, it's a good point. But it still adds controller, one is not exist before. And we have opportunity to remove it, but we cannot do it with RerouteController 🙈

Reply:
What it's true is that having setter and getter and a private property in Kotlin doesn't make too much sense to me, I believe `var rerouteController` should be enough.

For the other discussion let's see what others think. @kmadsen @korshaknn what's your preference? Any opinion?

Let's try to get an agreement here and unblock this PR from being merged 🚀

Reply:
+1 that we should have a consistent way to surface these APIs. 

The reason there is an attachArrivalController/removeArrivalController as a function instead of a property, is because it registerRouteProgress/unregisterRouteProgress. With a `var rerouteController` we'll also need a pattern to register/unregister. 

`createInternalOffRouteObserver`, `reroute`, and `interruptReroute` are all private functions needed for reroute but are still in `MapboxNavigation`. If those were moved inside `MapboxRerouteController`, it may change how the api should look. 

we can change arrival into a property, and add a `enableArrivalController(enable: Boolean)` function to handle the register/unregister. that might make the api more clear as well

Reply:
> but we cannot do it with RerouteController

Well, if we can do it for the arrival controller, we should be able to do it for the route controller as well. No controller, no automatic reroutes. I think that might actually be a desirable feature for some developers, especially the ones migrating from the legacy SDK since there was no automatic reroutes, they could just reuse their existing logic.

Reply:
I'd like to propose this:
```
var arrivalController: ArrivalController? = MapboxArrivalController
  set: {do the logic, also if `null`}
var rerouteController: RerouteController? = MapboxRerouteController
  set: {do the logic, also if `null`}
```
or
```
attachArrivalController
removeArrivalController
attachRouteController
getRerouteController
removeRerouteController
```
and do the same logic. Both should work similarly from the developer perspective - if the controller is `null`, don't handle the event (either reroute or arrival).

Reply:
@LukasPaczos @Guardiola31337  cannot make it via properties, because properties with custom setter/getter must be inited not via constructor, but at the same line. And of course it will work with functions 

@kmadsen yeah, it's a good point to move OffRouteObserver into RerouteController. I've tried it, it's not working clean enough because we force developers use offRouteObserver inside `RouteController` and we lose control when SDK can invoke reroute (might be another case when SDK would like to make reroute, I don't know scenarios but intuitively)

My main confuse here was: prevent use `attach*` for rerouteController because from my(developer) perspective _attach_ means we add smth new, but in current API we just replace one _rerouter_ by another. IMO as for developer it would be cleaner set/getRerouteController

PS @kmadsen I've made draft PR how it will look like for developer https://github.com/mapbox/mapbox-navigation-android/pull/3302

Reply:
```
setArrivalController(arrivalController: ArrivalController?)
setRouteController(routeController: RouteController?)
getRouteController() :RouteController?
```

looks good to me as well.

Reply:
@LukasPaczos like it 👍 

@kmadsen @Guardiola31337 folks thanks for debates, if you have any doubts, feel free to write it here. Also I'm gonna close #3302